File: code_files_MATLAB/LikelihoodFunc_Rate.m
function [sumloglik,logLik,rfTaylor,stdeps]=LikelihoodFunc_Rate(fedfund,inflation,outputgap,param)

n=length(fedfund);

rNbar=param(1,1);
betapi=abs(param(2,1));
betay=abs(param(3,1)); 
volr=abs(param(4,1)); %vol of residual. Not needed in the model

meaninflation=mean(inflation);
meanoutputgap=mean(outputgap);

for i=1:n,   
    rfTaylor(i,1)=rTaylor(rNbar,betapi,betay,inflation(i,1),outputgap(i,1),meaninflation,meanoutputgap);
    eps(i,1)=fedfund(i,1)-rfTaylor(i,1);
    vareps=volr^2;
    stdeps(i,1)=eps(i,1)/sqrt(vareps);
  
    % normal likelihood for whole time period
    epsVec=[eps(i,1)];
    VarMat=[vareps];
    Nv=length(epsVec);

    logLik(i,1)=log(1./((2.*pi)^(Nv/2).*sqrt(det(VarMat))))-1/2*(epsVec*(VarMat\epsVec'));
end


if abs(mean(rfTaylor)-mean(fedfund))>0.00005 || std(betapi*inflation)>4*std(betay*outputgap)
    pen=10^10;
else
    pen=0;
end
sumloglik=-sum(logLik(2:end,1))+pen; % this the sum of log likelihood (in negative sign since fminsearch minimizes)


end

File: code_files_R/LikelihoodFunc_Rate.R
LikelihoodFunc_Rate <- function(fedfund, inflation, outputgap, param) {
  # ---------------------------------------------------------------
  # LikelihoodFunc_Rate: Log-likelihood for Taylor Rule estimation
  # Input:
  #   fedfund   - observed nominal interest rate
  #   inflation - inflation rate series
  #   outputgap - output gap series
  #   param     - parameter vector (rNbar, betapi, betay, volr)
  # Output:
  #   sumloglik - total negative log-likelihood (for optimization)
  #   logLik    - vector of log-likelihood values
  #   rfTaylor  - Taylor rule implied interest rate
  #   stdeps    - standardized residuals
  # ---------------------------------------------------------------
  
  n <- length(fedfund)
  
  # Extract parameters
  rNbar <- param[1]
  betapi <- abs(param[2])
  betay <- abs(param[3])
  volr <- abs(param[4])   # standard deviation of residuals (not used directly in model)
  
  meaninflation <- mean(inflation)
  meanoutputgap <- mean(outputgap)
  
  # Initialize output vectors
  rfTaylor <- numeric(n)
  eps <- numeric(n)
  stdeps <- numeric(n)
  logLik <- numeric(n)
  
  for (i in seq_len(n)) {
    # Compute model-implied interest rate
    rfTaylor[i] <- rTaylor(
      rNbar, betapi, betay,
      inflation[i], outputgap[i],
      meaninflation, meanoutputgap
    )
    
    # Residuals
    eps[i] <- fedfund[i] - rfTaylor[i]
    vareps <- volr^2
    stdeps[i] <- eps[i] / sqrt(vareps)
    
    # Normal log-likelihood for each observation
    Nv <- 1
    logLik[i] <- log(1 / ((2 * pi)^(Nv / 2) * sqrt(vareps))) - 0.5 * (eps[i]^2 / vareps)
  }
  
  # Apply penalty if model misbehaves (same as MATLAB version)
  if (abs(mean(rfTaylor) - mean(fedfund)) > 0.00005 ||
      sd(betapi * inflation) > 4 * sd(betay * outputgap)) {
    pen <- 1e10
  } else {
    pen <- 0
  }
  
  # Sum of log-likelihood (negative for minimization)
  sumloglik <- -sum(logLik[-1]) + pen
  
  # Return results as list
  return(list(
    sumloglik = sumloglik,
    logLik = logLik,
    rfTaylor = rfTaylor,
    stdeps = stdeps
  ))
}
